os/board/rtl8721csm/src/component/os/tizenrt/tizenrt_service.c

```diff
@@ -34,26 +34,34 @@
  ******************************************************************************/
 extern void timer_wrapper(_timerHandle timer_hdl);
 
+struct _tizenrt_timer_entry {
+	struct list_head list;
+	struct timer_list_priv *timer;
+};
+
+static _list _tizenrt_timer_table;
+static u8 _tizenrt_timer_table_init = 0;
+
 static irqstate_t initial_tizen_flags, up_tizen_flag;
 static int flagcnt = 0;
 void save_and_cli()
 {
-       if(flagcnt){
-               up_tizen_flag = irqsave();
-       }else{
-               initial_tizen_flags = irqsave();
-       }
-       flagcnt++;
+	if(flagcnt){
+		up_tizen_flag = irqsave();
+	}else{
+		initial_tizen_flags = irqsave();
+	}
+	flagcnt++;
```

Comment:
Could you explain why did you use both up_tizen_flag and initial_tizen_flag? what is purpose of those value. 
please let me know to understand code better. Thanks in advance.

Reply:
It is for nested situation, first call of save_and_cli() will back up with up_tizen_flag.
The rest will be initial_tizen_flags, when restore_flags() is call only the last will restore with up_tizen_flag.

Reply:
Thanks for explanation. 
I checked tizenRT options and CONFIG_ARCH_HAVE_NESTED_INTERRUPT is related to this. 
Is there no possibility that CONFIG_ARCH_HAVE_NESTED_INTERRUPT  is disabled in tp1x? 
If so save_and_cli() code seems to be modified depend on the config. 

And second question is protection of flagcnt. 
I think two threads or interrupts can use up_tizen_flag. 
If IRQ B callled save_and_cli() before IRQ A increase flagcnt at line 55 then IRQ B uses up_tizen_flag which IRQ A used. 
Please let me know if I were wrong.

Reply:
It is not related to nested interrupt. Nested situation I mention is as below example. 
When Function A call save_and_cli() then after some execution is call Function B, however in Function B it call save_and_cli() again.

For second question, do you suggest that we use mute to protect?

Reply:
@ziyik, I cannot understand why you use two flags.
Could you explain in more detailed?
And even if we need two flags, there is a problem.
Let's assume the case like below.
Task A calls `save_and_cli()`, so tries to do `initial_tizen_flags = irqsave();`.
But before calling `irqsave()`, there is a context switching, so task B calls `save_and_cli()`.
In this case, also `initial_tizen_flag` will be used. (`flagcnt` will be increased to 1 from task B)
After returning to task A, it will use `initial_tizen_flag` again, and increases the `flagcnt` as 2.
And when calling `restore_flags()` from task A, it can successfully restore the irq. (decrease the `flagcnt` to 1)
But if task B tries to call `restore_flags()` again, `flagcnt` is still 1, so it tries to restore the irq with `up_tizen_flag`, not `initial_tizen_flags`.

Could you check this situation and elaborate in detail why we need two flags?

Reply:
The reason of 2 flag is for same Task calling save_and_cli() twice before restore_flags().
When Function A call save_and_cli() then after some execution is call Function B, however in Function B it call save_and_cli() again before restore_flags() is call, the the first flags will keep the initial setting, the second call of save_and_cli() setting will be store in up_tizen_flag.
When execute of Function B is done it will call restore_flags(), which restore the up_tizen_flag setting.
Then follow by Function A call restore_flags(), this time initial_tizen_flags will be restore.

The problem you mention only happen if nested interrupt can be happen, because when Function A call save_and_cli() interrupt have been disable and flagcnt will be increment and decrement after interrupt have been disable.

Reply:
CONFIG_ARCH_HAVE_NESTED_INTERRUPT=y is enabled in rtl8721csm. If the configuration enable nested interrupt then It means the issue joeengchan mentioned can happen.  Please check.

Reply:
@jeongchanKim @pillip8282 , the main purpose of this function is to prevent other task to interrupt.
Can I use sched_lock() and sched_unlock()?

Reply:
I used mutex for protection.
if sched_lock() and sched_unlock() can be use i will replace it with them.

Reply:
Sorry for late resposne. I've considered the issue cases that I menntioned. however I couldn't find appropriate solution to resolve it. 
@jeongchanKim please give your opinion about ziyik's solution.

Reply:
@ziyik I think you shouldn't use those HAL APIs. Then there is no problem as above. TizenRT irqsave and irqrestore support nested calling like
1. 1st irqsave
2. 2nd irqsave
3. 1st irqrestore for 2nd irqsave
4. 2nd irqrestore for 1st irqsave

Even if there is context swtching in the middle of steps above, there is no corruption for each other.

Reply:
@pillip8282 CONFIG_ARCH_HAVE_NESTED_INTERRUPT means this architecture can use nested interrupt, but it does not mean it's enabled. CONFIG_ARCH_NESTED_INTERRUPT means it. 
And this code is not related CONFIG_ARCH_NESTED_INTERRUPT. This is interrupt lock in SW. ARCH_NESTED_INTERRUPT is HW interrupt handling.

Reply:
@sunghan-chang , So I just need to take out the last commit will be fine is it?

Reply:
@ziyik No. If you keep your OSAL APIs, save_and_cli and restore_flags, it will have the issue as @jeongchanKim said. I suggested using TizenRT API for them. 
```
save_and_cli -> irqsave
restore_flags -> irqrestore
```
With that, you don't need to worry about nested case.

Reply:
How about the return value of irqsave() and pass in value of irqrestore()?
Caller of save_and_cli() and restore_flags() expect no return value and no pass in value.

Reply:
You need to use TizenRT way without your API. That's why I asked you before like below.
```
[sunghan-chang] [7 days ago] 
Are they OSAL APIs, save_and_cli and restore_flags? I mean, do you need to keep those functions?
```
If not, you need to save the flag for each thread separately. But I don't know how.